target/linux/ramips/image/mt7621.mk

```diff
@@ -630,6 +630,26 @@ define Device/gnubee_gb-pc2
 endef
 TARGET_DEVICES += gnubee_gb-pc2
 
+define Device/h3c_tx1806
+  $(Device/dsa-migration)
+  BLOCKSIZE := 128k
+  PAGESIZE := 2048
+  KERNEL_SIZE := 4096k
+  IMAGE_SIZE := 120832k
+  UBINIZE_OPTS := -E 5
+  KERNEL_LOADADDR := 0x82000000
+  KERNEL := kernel-bin | relocate-kernel 0x80001000 | lzma | \
+	fit lzma $$(KDIR)/image-$$(firstword $$(DEVICE_DTS)).dtb
+  IMAGES += factory.bin
+  IMAGE/sysupgrade.bin := sysupgrade-tar | append-metadata
+  IMAGE/factory.bin := pad-extra 160 | append-kernel | pad-to $$(KERNEL_SIZE) | \
```

Comment:
sysupgrade.bin not work due to u-boot load kernel from kenel@0xa0 (so that we need pad extra 160 bytes before kernel). What do I need to do to add offsets to sysupgrade.bin?

Reply:
Hmm, how about padding the kernel directly (in `KERNEL`)?
BTW why does it need extra 160 bytes? Is it some kind of vendor-defined header or just blank spaces?

Reply:
Yes, there should be an h3c header here, but it uses a non-standard crc32 algorithm, and I have no way of parsing it and no GPL code to refer to. This firmware header only affects upgrades from the web interface, u-boot does not verify this part.
H3C NX18 is a "SPI Flash Version" of H3C TX1806, just leave a firmware download link here, hope those who are experienced and interested can take a look.
[H3C NX18 Firmware](http://www.h3c.com/cn/d_202103/1389284_30005_0.htm)
some investigations:
0x00 ~ 0x07 fixed string "H3C_+k+f"
0x3c ~ 0x3f firmware version (time stamp or other, not sure)
0x44 ~ 0x47 symmetrical with “0x3c ~ 0x3f”
0x48 ~ 0x4f non-standard crc (kernel + rootfs I guess)
0x50 ~ 0x53 kernel size
0x54 ~ 0x57 rootfs size
0x62 ~ 0x70 firmware name (include board name)